---
title: "RNAseq_tutorial"
author: "Jin Cui"
date: "27/11/2018"
output: 
  html_document:
    toc: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=20, fig.height=8) 
```

# Description:

This tutorial contains data processing, visualization and differental expression analysis of single-end RNA-seq data. The RNA-seq count data comes from plant seedling tissues of maize natural mutant _Ufo1-1_ and wild-type (_Wt_) control.You can find additional information about _Ufo1-1_ from this [paper](http://www.plantcell.org/content/30/12/3006.abstract). I created my own transcriptome assembly using StringTie <https://ccb.jhu.edu/software/stringtie/>, hense the "MSTRG" gene_id in the count data. The reason I did that was because I not only want to study genes found in the reference genome, but also characterize new transcripts. I did pseudo-alignment and quantify transcript counts using Kallisto <https://pachterlab.github.io/kallisto/manual>. Gene counts are the sum of trancript pseudocounts. These counts are not in integers, so we need to round them first. Alternatively, you scan also generate raw count data using HTSeq <https://htseq.readthedocs.io/en/release_0.10.0/> or featurecounts <http://bioinf.wehi.edu.au/featureCounts/>. However, please do not use bedtools as it usually counts ambiguous alignments.


# Some additional reading on statistical analysis of RNA-seq:
<http://www.nathalievialaneix.eu/doc/pdf/tutorial-rnaseq.pdf>

```{r }
# pacman is a good package for loading packages
library(pacman)
pacman::p_load(edgeR, RColorBrewer, gplots, ggplot2, reshape2, DT, cowplot,
               limma, DESeq, DESeq2, data.table, e1071, ComplexHeatmap, VennDiagram)
```

# Source R script

I have writen several functions in an external R script, we can use those function by sourcing the function

```{r}
source("https://raw.githubusercontent.com/CJinnny/RNA-seq-tutorial/master/RNA_seq_tutorial_functions.R")

```

# Data preprocessing

In order to know gene functions, we can match MSTRG id into maize reference genome id (v4), which is already well characterized. Additionally, I also include maize reference genome V3 id.

```{r warning=FALSE}
require(data.table)
data <- fread("https://raw.githubusercontent.com/CJinnny/RNA-seq-tutorial/master/genes.txt", sep = "\t")
data[,6:15] = round(data[,6:15])

id_conv <- fread("https://raw.githubusercontent.com/CJinnny/RNA-seq-tutorial/master/https://raw.githubusercontent.com/CJinnny/RNA-seq-tutorial/master/all_MSTRG_v4_v3.csv", header = TRUE, sep = ",")
id_conv <- data.frame(id_conv, row.names = 1)
functions <- fread("https://raw.githubusercontent.com/CJinnny/RNA-seq-tutorial/master/B73v4_gene_function.txt", header = FALSE, sep = "\t")
functions <- subset(functions, select = c("V1", "V2"))
colnames(id_conv) <- c("v4_gene_id", "v3_gene_id", "MSTRG_gene_id")
colnames(functions) <- c("v4_gene_id", "gene_function")
annotations = merge(id_conv, functions, by = "v4_gene_id", all.x = TRUE)
data <- merge(data, annotations, by = "MSTRG_gene_id", all.x = TRUE)
data <- aggregate(data, 
                  by = list(data$MSTRG_gene_id),
                  FUN = aggregate_func
                  )
data <- na.omit(data, cols="v4_gene_id")
data <- data.frame(data, row.names = 1)
setnames(data, old = c("UU1", "UU2", "UU3", "UU4", "UU5", "WW1", "WW2", "WW3", "WW4", "WW5"), 
         new = c("U1", "U2","U3","U4","U5", "W1", "W2","W3","W4","W5"))


datatable(data, rownames = FALSE, width=0.5)

```

# Normality testing with skewness & kurtosis

```{r warning=FALSE}
counts <- data[,6:15]

require(e1071)
data_diagnosis(counts)
```

# Data transformation

Data is skewed (high skewness and high kurtosis).

We need to do data transformations for PCA plot, otherwise outliers will have a great impact on the clustering.

log transformation: logcounts = log2(counts + 1).

DESeq2::rlog(): "regularized log" transformation. For more information 
see <https://rdrr.io/bioc/DESeq2/man/rlog.html>

edgeR::cpm(): "counts per million" transformation. For more information 
see <https://rdrr.io/bioc/edgeR/man/cpm.html>

DESeq2:varianceStabilizingTransformation(): "variance stabilizing transformation". For more information see <https://rdrr.io/bioc/DESeq2/man/varianceStabilizingTransformation.html>

```{r message=FALSE}
require(DESeq2)
require(edgeR)
logcounts = log2(counts + 1)
rlogcounts = rlog(as.matrix(counts))
rownames(rlogcounts) = rownames(logcounts)
cpmcounts = cpm(as.matrix(counts), prior.count = 2, log = TRUE)
vstcounts = varianceStabilizingTransformation(as.matrix(counts))

```

```{r warning=FALSE}
data_diagnosis(logcounts)

```

```{r warning=FALSE}
data_diagnosis(rlogcounts)
```

```{r warning=FALSE}
data_diagnosis(cpmcounts)
```

```{r warning=FALSE}
data_diagnosis(vstcounts)
```

# Draw PCA (Principle Component Analysis) plot

Data normalization helps with separation

```{r fig1, fig.height = 12, fig.width = 18}
require(graphics)
require(RColorBrewer)
par(mfrow=c(2,3), mar=c(5.1, 4.6, 4.1, 1.6))
draw_PCA(counts, title = "PCA on raw data")
draw_PCA(logcounts, title = "PCA on log transformed data")
draw_PCA(rlogcounts, title = "PCA on rlog transformed data")
draw_PCA(cpmcounts, title = "PCA on cpm transformed data")
draw_PCA(vstcounts, title = "PCA on vst transformed data")
```

# Draw MDS (multidimensional scaling) plot

```{r fig2, fig.height = 12, fig.width = 18}
par(mfrow=c(2,3))
plotMDS(counts, col = c(rep("red", 5), rep("blue", 5)), cex = 1.5)
title("MDS plot on raw data")
plotMDS(logcounts, col = c(rep("red", 5), rep("blue", 5)), cex = 1.5)
title("MDS plot on log transformed data")
plotMDS(rlogcounts, col = c(rep("red", 5), rep("blue", 5)), cex = 1.5)
title("MDS plot on rlog transformed data")
plotMDS(cpmcounts, col = c(rep("red", 5), rep("blue", 5)), cex = 1.5)
title("MDS plot on cpm transformed data")
plotMDS(vstcounts, col = c(rep("red", 5), rep("blue", 5)), cex = 1.5)
title("MDS plot on vst transformed data")

```

# Draw correlation heatmap

```{r fig3, fig.height = 12, fig.width = 12}
require(gplots)
draw_corr_heatmap(as.matrix(counts), show_cellnote = TRUE, title = "clustering sample-to-sample\n distance on raw data")
draw_corr_heatmap(as.matrix(logcounts), show_cellnote = TRUE, title = "clustering sample-to-sample\n distance on log transformed data")
draw_corr_heatmap(rlogcounts, show_cellnote = TRUE, title = "clustering sample-to-sample\n distance on rlog transformed data")
draw_corr_heatmap(cpmcounts, show_cellnote = TRUE, title = "clustering sample-to-sample\n distance on cpm transformed data")
draw_corr_heatmap(vstcounts, show_cellnote = TRUE, title = "clustering sample-to-sample\n distance on vst transformed data")
```

# Set conditions for DE analysis

Here I have to relevel the conditions because by default it considers the first factor to be the reference

```{r}
conditions = factor(c(rep("Ufo",5), rep("Wt",5)))
conditions = relevel(conditions, ref="Wt")

```

# Filter low-count reads
I'm using cpm normalization method, filter out non-expressed genes (0 counts), you can also try more stringent filtering (e.g. cpm > 1 in at least 5 samples)

```{r}
keep <- rowSums(cpm(counts)>0) >= 5
table(keep)
keep_true = data.frame(keep[which(keep == TRUE)])
filter = subset(counts, rownames(counts) %in% rownames(keep_true))
```

# DESeq

```{r}
require(DESeq)
cds = newCountDataSet(filter, conditions)
cds = DESeq::estimateSizeFactors(cds)
cds = DESeq::estimateDispersions(cds)
DESeq_res = DESeq::nbinomTest(cds, "Wt", "Ufo")
rownames(DESeq_res) = DESeq_res$id
DESeq_DE = subset(DESeq_res, (log2FoldChange < -1 & padj < 0.05) | (log2FoldChange > 1 & padj < 0.05) )
DESeq_nc = counts(cds, normalized = TRUE)
DESeq_nc = data.frame("id" = rownames(DESeq_nc), DESeq_nc)
DESeq_nc_DE = subset(DESeq_nc, id %in% DESeq_DE$id)
```

# DESeq2

```{r}
require(DESeq2)
colData = data.frame(samples = colnames(filter), conditions = conditions)
dds = DESeqDataSetFromMatrix(countData = filter, colData = colData, design = ~conditions)
dds = DESeq2::estimateSizeFactors(dds)
dds = DESeq2::estimateDispersions(dds)
dds = DESeq2::nbinomWaldTest(dds)

DESeq2_res = DESeq2::results(dds)
DESeq2_res = data.frame("id" = rownames(DESeq2_res), DESeq2_res)
DESeq2_DE = subset(DESeq2_res, (log2FoldChange < -1 & padj < 0.05) | (log2FoldChange > 1 & padj < 0.05) )
DESeq2_nc = counts(dds, normalized = TRUE)
DESeq2_nc = data.frame("id" = rownames(DESeq2_nc), DESeq2_nc)
DESeq2_nc_DE = subset(DESeq2_nc, id %in% DESeq2_DE$id)
```

# edgeR

```{r}
require(edgeR)
group = as.vector(conditions)
dge = DGEList(counts = filter, group = group)
dge = calcNormFactors(dge)
dis = estimateCommonDisp(dge)
tag = estimateTagwiseDisp(dis)
etx = exactTest(tag, pair = c("Wt", "Ufo"))
edgeR_res = etx$table
edgeR_res$FDR = p.adjust(edgeR_res$PValue, method = "BH")
edgeR_res = data.frame("id" = rownames(edgeR_res), edgeR_res)
edgeR_DE = subset(edgeR_res, (logFC < -1 & FDR < 0.05) | (logFC > 1 & FDR < 0.05) )
edgeR_nc = tag$pseudo.counts
edgeR_nc = data.frame("id"=rownames(edgeR_nc), edgeR_nc)
edgeR_nc_DE = subset(edgeR_nc, id %in% edgeR_DE$id)
```

# limma

```{r}
require(limma)
design = model.matrix(~conditions)
voom = voom(filter, design, normalize="quantile")
fit = lmFit(voom, design)
fit = eBayes(fit)
limma_res = topTable(fit, coef = NULL, n = Inf)
limma_res = data.frame("id" = rownames(limma_res), limma_res)
limma_DE = subset(limma_res, (logFC < -1 & adj.P.Val < 0.05) | (logFC > 1 & adj.P.Val < 0.05) )
limma_nc = 2**voom$E
limma_nc = data.frame("id" = rownames(limma_nc), limma_nc)
limma_nc_DE = subset(limma_nc, id %in% limma_DE$id)
```

# Compare DESeq, DESeq2, edgeR and limma DEG results

```{r}
dflist <- list(DESeq=DESeq_DE, DESeq2=DESeq2_DE, edgeR=edgeR_DE, limma=limma_DE)
Compare_method <- join_id(dflist)
Compare_method$message
```

# Displaying table using DT library

```{r warning=FALSE}
require(DT)
Compare_summary <- merge(Compare_method$merged_table, annotations, by.x = "id", by.y = "MSTRG_gene_id", all.x = TRUE)
datatable(Compare_summary,
          rownames = FALSE,
          caption = htmltools::tags$caption(
            style = "caption-side: bottom; text-align: left;",
            "Table 1: Summary of DEGs found in four different methods."),
          options = list(
            searchHighlight = TRUE
          )
          )
```

# Add log2FoldChange

```{r warning=FALSE}
Compare_summary <- merge(Compare_summary, DESeq_res[,c("id", "log2FoldChange","padj")], by = "id")
setnames(Compare_summary, old = c("log2FoldChange", "padj"), new = c("DESeq_logFC", "DESeq_padj"))
Compare_summary <- merge(Compare_summary, DESeq2_res[,c("id", "log2FoldChange","padj")], by = "id")
setnames(Compare_summary, old = c("log2FoldChange", "padj"), new = c("DESeq2_logFC", "DESeq2_padj"))
Compare_summary <- merge(Compare_summary, edgeR_res[,c("id", "logFC", "FDR")], by = "id")
setnames(Compare_summary, old = c("logFC", "FDR"), new = c("edgeR_logFC", "edgeR_padj"))
Compare_summary <- merge(Compare_summary, limma_res[,c("id", "logFC", "adj.P.Val")], by = "id")
setnames(Compare_summary, old = c("logFC", "adj.P.Val"), new = c("limma_logFC", "limma_padj"))
Compare_summary = Compare_summary[,c(1,9,10,11,12,13,14,15,16,6,7,8)]
Compare_summary = aggregate(Compare_summary, by = list(Compare_summary$id), FUN = aggregate_func)
Compare_summary$id <- NULL
setnames(Compare_summary, old = "Group.1", new = "id")
datatable(Compare_summary, filter = "top",
          rownames = FALSE,
          caption = htmltools::tags$caption(
            style = "caption-side: bottom; text-align: left;",
            "Table 2: Summary of logFC values of DEGs."),
          options = list(
            autoWidth = TRUE,
            searchHighlight = TRUE,
            scrollX = TRUE
            )
          )
```

# Draw Venn Diagram

```{r fig.height = 12, fig.width = 12}
require(VennDiagram)
draw_venndiagram(dflist, Compare_method$merged_table)
```


# Draw heatmap for DEGs in separate plots

Cluster for Up & Down DEGs, respectively

```{r fig4, fig.height = 12, fig.width = 12}
require(ComplexHeatmap)
Ht1 = DE_heatmap(DESeq_nc_DE, title = "DESeq", km = 2)
Ht2 = DE_heatmap(DESeq2_nc_DE, title = "DESeq2", km = 2)
Ht3 = DE_heatmap(edgeR_nc_DE, title = "edgeR", km = 2)
Ht4 = DE_heatmap(limma_nc_DE, title = "limma", km = 2)
Ht1
Ht2
Ht3
Ht4
```

# Draw heatmaps side-by-side on common DEGs

```{r fig5, fig.height = 8, fig.width = 18, warning = FALSE}
Ht1 = DE_heatmap(DESeq_nc_DE, common_id=Compare_method$common_id, title = "DESeq", km = 2)
Ht2 = DE_heatmap(DESeq2_nc_DE,common_id=Compare_method$common_id, title = "DESeq2",km = 2)
Ht3 = DE_heatmap(edgeR_nc_DE, common_id=Compare_method$common_id, title = "edgeR", km = 2)
Ht4 = DE_heatmap(limma_nc_DE, common_id=Compare_method$common_id, title = "limma", km = 2)

Ht1 + Ht2 + Ht3 + Ht4
```

# Draw MA plot

```{r fig6, fig.height = 18, fig.width = 18}
par(mfrow=c(2,2), mar=c(5.1, 4.6, 4.1, 1.6))
draw_MA(DESeq_res, type="DESeq")
draw_MA(DESeq2_res, type="DESeq2")
draw_MA(edgeR_res, type="edgeR")
draw_MA(limma_res, type="limma")
```

# Draw Volcano plot

```{r fig7, fig.height = 18, fig.width = 18}
par(mfrow=c(2,2), mar=c(5.1, 4.6, 4.1, 1.6))
draw_volcano(DESeq_res, type="DESeq")
draw_volcano(DESeq2_res, type="DESeq2")
draw_volcano(edgeR_res, type="edgeR")
draw_volcano(limma_res, type="limma")
```

```{r}
sessionInfo()
```